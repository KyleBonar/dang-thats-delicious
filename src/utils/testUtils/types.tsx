export interface MockLink {
  text: string;
  to: string;
  target: "_blank" | "_self";
}
