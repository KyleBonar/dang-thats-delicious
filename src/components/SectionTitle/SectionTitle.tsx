import * as React from "react";
import styled from "../../theme/styled-components";

const StyledDiv = styled.div`
  width: 100%;
  text-align: center;
`;

interface SectionTitleProps {
  title: string;
  description: string;
}

const SectionTitle: React.SFC<SectionTitleProps> = props => {
  return (
    <StyledDiv>
      <h2>{props.title}</h2>
      <h6>{props.description}</h6>
    </StyledDiv>
  );
};

export default SectionTitle;
