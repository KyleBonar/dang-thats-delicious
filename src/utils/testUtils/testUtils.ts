import * as React from "react";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";

import { MockCard } from "./types";
import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import { FlashMessageProps } from "../../components/FlashMessage/FlashMessage";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.string,
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.array_of_words(),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = new Array(casual.integer(1, 15))
    .fill(null)
    .map(fakeSauce);
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured: undefined,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const storeConfig = configureStore([]);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const fakeCard = (): MockCard => ({
  title: casual.string,
  description: casual.string,
  to: casual.url,
  showLink: casual.boolean,
  imageLink: casual.url,
  className: casual.string,
  anchorText: casual.string
});

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(0, MIN_PASSWORD_LENGTH);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword
};
