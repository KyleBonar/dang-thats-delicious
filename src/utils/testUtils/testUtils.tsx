import * as React from "react";
import * as enzyme from "enzyme";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";
import { IErrReturn } from "../Err/Err";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.random_element([undefined, casual.string]),
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.random_element([undefined, casual.array_of_words()]),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const generateFakeSauces = (): ISauce[] =>
  new Array(casual.integer(1, 15)).fill(null).map(fakeSauce);

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = generateFakeSauces();
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  const featured: undefined | string[] = casual.random_element([
    undefined,
    allSlugs.length > 0
      ? allSlugs.filter(slug => {
          if (casual.random < 0.4) return slug;
          return;
        })
      : undefined
  ]);

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const middlewares = [thunk];
  const storeConfig = configureStore(middlewares);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(1, MIN_PASSWORD_LENGTH - 1);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

// Function designed to resemble how a network error would look like.
// This will help the catch(err) only have to handle a single format
const generateErr = (): IErrReturn => {
  return {
    response: {
      data: {
        status: 400,
        msg: casual.string,
        isGood: false
      }
    }
  };
};

// lets us mount up a simple custom hook
const mountReactHook = hook => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.shallow(
      <Component>
        {hookValues => {
          Object.assign(componentHook, hookValues);
          return null;
        }}
      </Component>
    );
  });
  return { componentMount, componentHook };
};

export interface ImountReactHookWithReduxStore {
  componentMount: any;
  componentHook: any;
}
const mountReactHookWithReduxStore = (
  hook,
  store
): ImountReactHookWithReduxStore => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.mount(
      <Provider store={store}>
        <Component>
          {hookValues => {
            Object.assign(componentHook, hookValues);
            return null;
          }}
        </Component>
      </Provider>
    );
  });
  return { componentMount, componentHook };
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword,
  generateInValidPassword,
  generateErr,
  mountReactHook,
  mountReactHookWithReduxStore
};
