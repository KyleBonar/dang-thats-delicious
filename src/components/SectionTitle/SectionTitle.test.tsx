import * as React from "react";
import * as enzyme from "enzyme";

import SectionTitle, { SectionTitleProps } from "./SectionTitle";
import { casual, ITERATION_SIZE } from "../../utils/testUtils/testUtils";

const fakeSectionTitle = (): SectionTitleProps => ({
  title: casual.string,
  description: casual.string
});

const mockSectionTitles = new Array(ITERATION_SIZE)
  .fill(null)
  .map(fakeSectionTitle);

describe("<SectionTitle />", () => {
  it("renders", () => {
    mockSectionTitles.forEach(mockSectionTitle => {
      const wrapper = enzyme.shallow(<SectionTitle {...mockSectionTitle} />);

      expect(wrapper).toBeTruthy();
    });
  });

  it("renders correct tags", () => {
    mockSectionTitles.forEach(mockSectionTitle => {
      const wrapper = enzyme.shallow(<SectionTitle {...mockSectionTitle} />);

      expect(wrapper.find("h2")).toBeTruthy();
      expect(wrapper.find("h2").length).toEqual(1);

      expect(wrapper.find("h6")).toBeTruthy();
      expect(wrapper.find("h6").length).toEqual(1);
    });
  });

  it("renders correct h2 text", () => {
    mockSectionTitles.forEach(mockSectionTitle => {
      const wrapper = enzyme.shallow(<SectionTitle {...mockSectionTitle} />);

      expect(wrapper.find("h2").render().text()).toEqual(
        mockSectionTitle.title
      );
    });
  });

  it("renders correct h6 text", () => {
    mockSectionTitles.forEach(mockSectionTitle => {
      const wrapper = enzyme.shallow(<SectionTitle {...mockSectionTitle} />);

      expect(wrapper.find("h6").render().text()).toEqual(
        mockSectionTitle.description
      );
    });
  });
});
