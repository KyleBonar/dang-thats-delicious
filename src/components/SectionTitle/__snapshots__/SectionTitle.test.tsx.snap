// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SectionTitle /> matches snapshot 1`] = `
<div>
  <h2>
    sequi facilis corporis sed veritatis quidem praesentium
  </h2>
  <h6>
    magnam fugiat error molestiae laborum est possimus
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 2`] = `
<div>
  <h2>
    assumenda aut enim iure facilis ut explicabo
  </h2>
  <h6>
    iure vitae occaecati officia iure nisi ut
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 3`] = `
<div>
  <h2>
    maiores recusandae fuga reprehenderit soluta non omnis
  </h2>
  <h6>
    consequatur enim quia autem et modi tenetur
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 4`] = `
<div>
  <h2>
    amet modi reprehenderit omnis adipisci eius laborum
  </h2>
  <h6>
    eius aut illo dolorum ab aliquam id
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 5`] = `
<div>
  <h2>
    sit delectus vel ipsa tempora aut sint
  </h2>
  <h6>
    aperiam occaecati illum animi vel provident dicta
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 6`] = `
<div>
  <h2>
    perferendis dolorum cupiditate illum saepe ut assumenda
  </h2>
  <h6>
    sint quia sed exercitationem omnis laborum exercitationem
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 7`] = `
<div>
  <h2>
    est sapiente laudantium molestias assumenda quasi adipisci
  </h2>
  <h6>
    mollitia iste itaque omnis debitis rem similique
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 8`] = `
<div>
  <h2>
    temporibus aliquam sed sapiente soluta corporis commodi
  </h2>
  <h6>
    eaque nisi cum ducimus nemo doloremque maiores
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 9`] = `
<div>
  <h2>
    esse atque et non explicabo dolores ipsam
  </h2>
  <h6>
    laboriosam non itaque voluptas ea ducimus autem
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 10`] = `
<div>
  <h2>
    sit aspernatur aspernatur quam eligendi est fugiat
  </h2>
  <h6>
    dolorem nihil quam nulla est unde sit
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 11`] = `
<div>
  <h2>
    ex expedita est doloremque explicabo eaque necessitatibus
  </h2>
  <h6>
    sit suscipit omnis necessitatibus maiores ullam quae
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 12`] = `
<div>
  <h2>
    corporis qui aut nobis quo consequatur harum
  </h2>
  <h6>
    qui enim et doloremque voluptate in sunt
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 13`] = `
<div>
  <h2>
    occaecati nulla aut sunt est rerum at
  </h2>
  <h6>
    sint dolores fugiat quis et voluptas culpa
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 14`] = `
<div>
  <h2>
    et vero repellat eveniet assumenda quis quae
  </h2>
  <h6>
    iusto sapiente nostrum aut corrupti sed consectetur
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 15`] = `
<div>
  <h2>
    soluta inventore laudantium dolor sint officiis qui
  </h2>
  <h6>
    dignissimos dolore qui voluptatem cumque nihil iste
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 16`] = `
<div>
  <h2>
    est unde rem similique et maxime quia
  </h2>
  <h6>
    dolores quisquam aut consequatur sint provident aut
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 17`] = `
<div>
  <h2>
    ullam distinctio autem quam qui facere in
  </h2>
  <h6>
    ut odit animi in ut occaecati at
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 18`] = `
<div>
  <h2>
    odit tempora at modi occaecati illum laborum
  </h2>
  <h6>
    quos eum voluptate sit saepe assumenda deserunt
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 19`] = `
<div>
  <h2>
    consequatur itaque et in fugit in rerum
  </h2>
  <h6>
    libero odit aut labore qui aut inventore
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 20`] = `
<div>
  <h2>
    aut eveniet beatae incidunt ut ducimus fugiat
  </h2>
  <h6>
    et et aut quidem voluptatibus libero aut
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 21`] = `
<div>
  <h2>
    autem ut ullam et possimus sint fugiat
  </h2>
  <h6>
    maiores autem est esse quos minus deserunt
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 22`] = `
<div>
  <h2>
    nihil quasi minus quo sunt possimus earum
  </h2>
  <h6>
    animi quis expedita et optio consequatur velit
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 23`] = `
<div>
  <h2>
    voluptate necessitatibus excepturi et qui sunt nesciunt
  </h2>
  <h6>
    provident eaque atque quos rerum nobis nam
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 24`] = `
<div>
  <h2>
    dolor quas repellendus in dolores totam animi
  </h2>
  <h6>
    ipsum occaecati et consequatur consectetur sunt molestiae
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 25`] = `
<div>
  <h2>
    cupiditate autem consequatur consequatur accusantium molestiae non
  </h2>
  <h6>
    sint et ipsa consequuntur quidem dolorem quia
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 26`] = `
<div>
  <h2>
    consequatur omnis ab assumenda animi non quis
  </h2>
  <h6>
    quis nobis repudiandae aliquam molestiae occaecati a
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 27`] = `
<div>
  <h2>
    culpa facilis eligendi voluptatibus similique blanditiis in
  </h2>
  <h6>
    eum ut dignissimos aspernatur voluptate voluptatem est
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 28`] = `
<div>
  <h2>
    eius dolorem amet quia quia aut sint
  </h2>
  <h6>
    cumque sunt mollitia quibusdam officia eum repudiandae
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 29`] = `
<div>
  <h2>
    minima temporibus ducimus quia assumenda ut neque
  </h2>
  <h6>
    saepe est labore sed rerum nesciunt voluptas
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 30`] = `
<div>
  <h2>
    rem repellendus nihil repudiandae aliquid ex architecto
  </h2>
  <h6>
    ut corrupti voluptatem quia dolores dolor voluptates
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 31`] = `
<div>
  <h2>
    sequi perferendis nostrum aliquid ea animi architecto
  </h2>
  <h6>
    accusantium vel enim sint voluptatem porro ut
  </h6>
</div>
`;

exports[`<SectionTitle /> matches snapshot 32`] = `
<div>
  <h2>
    est sed qui repellat assumenda et incidunt
  </h2>
  <h6>
    aut distinctio molestias autem expedita omnis sit
  </h6>
</div>
`;
